File: web/src/lib/timeline-parser.ts
/**
 * Parser for the .timeline format.
 * Inverse of timeline-serializer.ts.
 * Converts .timeline text → ParsedTimeline object.
 */

export interface ParsedTask {
	id: string | null;
	name: string;
	status: string;
	color: string | null;
	planned_start: string | null;
	planned_end: string | null;
	actual_start: string | null;
	actual_end: string | null;
	blocked_reason: string | null;
	note: string | null;
	url: string | null;
	sort_order: number;
	depends_on: string[];
}

export interface ParsedSection {
	id: string | null;
	name: string;
	sort_order: number;
	tasks: ParsedTask[];
}

export interface ParsedMilestone {
	id: string | null;
	date: string | null;
	label: string;
	kind: "milestone" | "phase";
	sort_order: number;
}

export interface ParsedAnnouncement {
	id: string | null;
	date: string | null;
	content: string;
	type: "note" | "downtime" | "general";
}

export interface ParsedTimeline {
	title: string;
	sections: ParsedSection[];
	milestones: ParsedMilestone[];
	announcements: ParsedAnnouncement[];
}

// Extract embedded ID comment: <!-- type:uuid -->
const ID_RE = /\s*<!--\s*(\w+):([\w-]+)\s*-->$/;

function extractId(line: string): { clean: string; id: string | null } {
	const match = line.match(ID_RE);
	if (match) {
		return { clean: line.slice(0, match.index!).trimEnd(), id: match[2]! };
	}
	return { clean: line, id: null };
}

// Extract quoted string: "content here"
function extractQuoted(s: string): string {
	const match = s.match(/"([^"]*)"/);
	return match ? match[1]! : s.trim();
}

const DATE_RE = /\d{4}-\d{2}-\d{2}/;
const TASK_RE = /^- \[([^\]]+)\]\s+(.+)/;
const DATE_RANGE_RE = /(\d{4}-\d{2}-\d{2})\s*->\s*(\d{4}-\d{2}-\d{2})/;
const COLOR_RE = /#[0-9A-Fa-f]{6}\b/;

export function parseTimeline(text: string): ParsedTimeline {
	const lines = text.split("\n");

	const result: ParsedTimeline = {
		title: "",
		sections: [],
		milestones: [],
		announcements: [],
	};

	let inFrontmatter = false;
	let context: "none" | "phases" | "milestones" | "notes" | "timeline" = "none";
	let currentSection: ParsedSection | null = null;
	let currentTask: ParsedTask | null = null;
	let milestoneOrder = 0;
	let sectionOrder = 0;
	let taskOrder = 0;

	for (const rawLine of lines) {
		const trimmed = rawLine.trim();

		// Skip empty lines
		if (trimmed === "") {
			currentTask = null;
			continue;
		}

		// Frontmatter
		if (trimmed === "---") {
			inFrontmatter = !inFrontmatter;
			continue;
		}

		if (inFrontmatter) {
			const colonIdx = trimmed.indexOf(":");
			if (colonIdx !== -1) {
				const key = trimmed.slice(0, colonIdx).trim();
				const value = trimmed.slice(colonIdx + 1).trim();
				if (key === "title") {
					result.title = value;
				}
			}
			continue;
		}

		// Headings
		if (trimmed.startsWith("# ") && !trimmed.startsWith("## ")) {
			const heading = trimmed.slice(2).trim().toLowerCase();
			if (heading === "phases") context = "phases";
			else if (heading === "milestones") context = "milestones";
			else if (heading === "notes") context = "notes";
			else if (heading === "timeline") context = "timeline";
			continue;
		}

		// Section header
		if (trimmed.startsWith("## ")) {
			const { clean, id } = extractId(trimmed);
			const name = clean.slice(3).trim();
			currentSection = {
				id,
				name,
				sort_order: sectionOrder++,
				tasks: [],
			};
			result.sections.push(currentSection);
			currentTask = null;
			taskOrder = 0;
			context = "timeline";
			continue;
		}

		// Phase line: ~ DATE "label"
		if (trimmed.startsWith("~ ")) {
			const { clean, id } = extractId(trimmed);
			const rest = clean.slice(2).trim();
			const dateMatch = rest.match(DATE_RE);
			const label = extractQuoted(rest);
			result.milestones.push({
				id,
				date: dateMatch ? dateMatch[0] : null,
				label,
				kind: "phase",
				sort_order: milestoneOrder++,
			});
			continue;
		}

		// Milestone line: @ DATE "label"
		if (trimmed.startsWith("@ ")) {
			const { clean, id } = extractId(trimmed);
			const rest = clean.slice(2).trim();
			const dateMatch = rest.match(DATE_RE);
			const label = extractQuoted(rest);
			result.milestones.push({
				id,
				date: dateMatch ? dateMatch[0] : null,
				label,
				kind: "milestone",
				sort_order: milestoneOrder++,
			});
			continue;
		}

		// Announcement line: ! DATE [type] "content"
		if (trimmed.startsWith("! ")) {
			const { clean, id } = extractId(trimmed);
			const rest = clean.slice(2).trim();
			const dateMatch = rest.match(DATE_RE);
			const typeMatch = rest.match(/\[(\w+)\]/);
			const content = extractQuoted(rest);
			let type: "note" | "downtime" | "general" = "general";
			if (typeMatch) {
				const t = typeMatch[1]!.toLowerCase();
				if (t === "note" || t === "downtime") type = t;
			}
			result.announcements.push({
				id,
				date: dateMatch ? dateMatch[0] : null,
				content,
				type,
			});
			continue;
		}

		// Task line: - [status] name | start -> end | color
		const taskMatch = trimmed.match(TASK_RE);
		if (taskMatch && currentSection) {
			const { clean, id } = extractId(trimmed);
			const fullTaskMatch = clean.match(TASK_RE);
			if (!fullTaskMatch) continue;

			const status = fullTaskMatch[1]!;
			const restOfLine = fullTaskMatch[2]!;

			// Split by pipes
			const parts = restOfLine.split("|").map((p) => p.trim());
			const name = parts[0]!;

			let planned_start: string | null = null;
			let planned_end: string | null = null;
			let color: string | null = null;

			for (let i = 1; i < parts.length; i++) {
				const part = parts[i]!;
				const rangeMatch = part.match(DATE_RANGE_RE);
				if (rangeMatch) {
					planned_start = rangeMatch[1]!;
					planned_end = rangeMatch[2]!;
				} else {
					const colorMatch = part.match(COLOR_RE);
					if (colorMatch) {
						color = colorMatch[0];
					}
				}
			}

			currentTask = {
				id,
				name,
				status,
				color,
				planned_start,
				planned_end,
				actual_start: null,
				actual_end: null,
				blocked_reason: null,
				note: null,
				url: null,
				sort_order: taskOrder++,
				depends_on: [],
			};
			currentSection.tasks.push(currentTask);
			continue;
		}

		// Metadata lines (indented, belong to current task)
		if (rawLine.startsWith("  ") && currentTask) {
			const metaLine = trimmed;
			const colonIdx = metaLine.indexOf(":");
			if (colonIdx === -1) continue;

			const key = metaLine.slice(0, colonIdx).trim();
			const value = metaLine.slice(colonIdx + 1).trim();

			switch (key) {
				case "actual": {
					const rangeMatch = value.match(DATE_RANGE_RE);
					if (rangeMatch) {
						currentTask.actual_start = rangeMatch[1]!;
						currentTask.actual_end = rangeMatch[2]!;
					}
					break;
				}
				case "actual-start": {
					const dateMatch = value.match(DATE_RE);
					if (dateMatch) currentTask.actual_start = dateMatch[0];
					break;
				}
				case "actual-end": {
					const dateMatch = value.match(DATE_RE);
					if (dateMatch) currentTask.actual_end = dateMatch[0];
					break;
				}
				case "blocked-by":
					currentTask.blocked_reason = extractQuoted(value);
					break;
				case "note":
					currentTask.note = extractQuoted(value);
					break;
				case "url":
					currentTask.url = value;
					break;
				case "depends":
					currentTask.depends_on = value.split(",").map((d) => d.trim()).filter(Boolean);
					break;
			}
			continue;
		}
	}

	return result;
}

// --- Validation ---

const VALID_STATUSES = new Set(["todo", "in-progress", "done", "blocked", "cancelled"]);

interface ValidationResult {
	valid: boolean;
	errors: string[];
	parsed: ParsedTimeline | null;
}

export function validateTimeline(text: string): ValidationResult {
	const errors: string[] = [];
	const parsed = parseTimeline(text);

	if (!parsed.title || !parsed.title.trim()) {
		errors.push("Title is required");
	}

	if (parsed.sections.length === 0) {
		errors.push("At least one section is required");
	}

	for (const section of parsed.sections) {
		for (const task of section.tasks) {
			if (!VALID_STATUSES.has(task.status)) {
				errors.push(`Task "${task.name}": invalid status "${task.status}" (must be one of: ${[...VALID_STATUSES].join(", ")})`);
			}

			if (task.planned_start && task.planned_end) {
				if (task.planned_end < task.planned_start) {
					errors.push(`Task "${task.name}": end date (${task.planned_end}) is before start date (${task.planned_start})`);
				}
			}
		}
	}

	return {
		valid: errors.length === 0,
		errors,
		parsed: errors.length === 0 ? parsed : null,
	};
}

File: server/app/core/timeline_parser.py
"""
Parser for the .timeline format.
Python port of web/src/lib/timeline-parser.ts.
Converts .timeline text → dict structure for JSON response.
"""

import re
from typing import Any

ID_RE = re.compile(r"\s*<!--\s*(\w+):([\w-]+)\s*-->$")
DATE_RE = re.compile(r"\d{4}-\d{2}-\d{2}")
TASK_RE = re.compile(r"^- \[([^\]]+)\]\s+(.+)")
DATE_RANGE_RE = re.compile(r"(\d{4}-\d{2}-\d{2})\s*->\s*(\d{4}-\d{2}-\d{2})")
COLOR_RE = re.compile(r"#[0-9A-Fa-f]{6}\b")


def _extract_id(line: str) -> tuple[str, str | None]:
    """Extract and strip embedded ID comment from end of line."""
    m = ID_RE.search(line)
    if m:
        return line[: m.start()].rstrip(), m.group(2)
    return line, None


def _extract_quoted(s: str) -> str:
    """Extract content from first quoted string."""
    m = re.search(r'"([^"]*)"', s)
    return m.group(1) if m else s.strip()


def parse_timeline(text: str) -> dict[str, Any]:
    """Parse .timeline text into a structured dict."""
    result: dict[str, Any] = {
        "title": "",
        "sections": [],
        "milestones": [],
        "announcements": [],
    }

    in_frontmatter = False
    current_section: dict | None = None
    current_task: dict | None = None
    milestone_order = 0
    section_order = 0
    task_order = 0

    for raw_line in text.split("\n"):
        trimmed = raw_line.strip()

        if not trimmed:
            current_task = None
            continue

        # Frontmatter
        if trimmed == "---":
            in_frontmatter = not in_frontmatter
            continue

        if in_frontmatter:
            colon = trimmed.find(":")
            if colon != -1:
                key = trimmed[:colon].strip()
                value = trimmed[colon + 1 :].strip()
                if key == "title":
                    result["title"] = value
            continue

        # Headings (context markers — we don't need to track context for parsing)
        if trimmed.startswith("# ") and not trimmed.startswith("## "):
            continue

        # Section header
        if trimmed.startswith("## "):
            clean, sid = _extract_id(trimmed)
            name = clean[3:].strip()
            current_section = {
                "id": sid,
                "name": name,
                "sort_order": section_order,
                "tasks": [],
            }
            section_order += 1
            result["sections"].append(current_section)
            current_task = None
            task_order = 0
            continue

        # Phase
        if trimmed.startswith("~ "):
            clean, mid = _extract_id(trimmed)
            rest = clean[2:].strip()
            date_match = DATE_RE.search(rest)
            label = _extract_quoted(rest)
            result["milestones"].append({
                "id": mid,
                "date": date_match.group(0) if date_match else None,
                "label": label,
                "kind": "phase",
                "sort_order": milestone_order,
            })
            milestone_order += 1
            continue

        # Milestone
        if trimmed.startswith("@ "):
            clean, mid = _extract_id(trimmed)
            rest = clean[2:].strip()
            date_match = DATE_RE.search(rest)
            label = _extract_quoted(rest)
            result["milestones"].append({
                "id": mid,
                "date": date_match.group(0) if date_match else None,
                "label": label,
                "kind": "milestone",
                "sort_order": milestone_order,
            })
            milestone_order += 1
            continue

        # Announcement
        if trimmed.startswith("! "):
            clean, aid = _extract_id(trimmed)
            rest = clean[2:].strip()
            date_match = DATE_RE.search(rest)
            type_match = re.search(r"\[(\w+)\]", rest)
            content = _extract_quoted(rest)
            ann_type = "general"
            if type_match:
                t = type_match.group(1).lower()
                if t in ("note", "downtime"):
                    ann_type = t
            result["announcements"].append({
                "id": aid,
                "date": date_match.group(0) if date_match else None,
                "content": content,
                "type": ann_type,
            })
            continue

        # Task line
        task_match = TASK_RE.match(trimmed)
        if task_match and current_section is not None:
            clean, tid = _extract_id(trimmed)
            full_match = TASK_RE.match(clean)
            if not full_match:
                continue

            status = full_match.group(1)
            rest_of_line = full_match.group(2)
            parts = [p.strip() for p in rest_of_line.split("|")]
            name = parts[0]

            planned_start = None
            planned_end = None
            color = None

            for part in parts[1:]:
                range_match = DATE_RANGE_RE.search(part)
                if range_match:
                    planned_start = range_match.group(1)
                    planned_end = range_match.group(2)
                else:
                    color_match = COLOR_RE.search(part)
                    if color_match:
                        color = color_match.group(0)

            current_task = {
                "id": tid,
                "name": name,
                "status": status,
                "color": color,
                "planned_start": planned_start,
                "planned_end": planned_end,
                "actual_start": None,
                "actual_end": None,
                "blocked_reason": None,
                "note": None,
                "url": None,
                "sort_order": task_order,
                "depends_on": [],
            }
            task_order += 1
            current_section["tasks"].append(current_task)
            continue

        # Metadata lines (indented)
        if raw_line.startswith("  ") and current_task is not None:
            colon = trimmed.find(":")
            if colon == -1:
                continue
            key = trimmed[:colon].strip()
            value = trimmed[colon + 1 :].strip()

            if key == "actual":
                range_match = DATE_RANGE_RE.search(value)
                if range_match:
                    current_task["actual_start"] = range_match.group(1)
                    current_task["actual_end"] = range_match.group(2)
            elif key == "actual-start":
                date_match = DATE_RE.search(value)
                if date_match:
                    current_task["actual_start"] = date_match.group(0)
            elif key == "actual-end":
                date_match = DATE_RE.search(value)
                if date_match:
                    current_task["actual_end"] = date_match.group(0)
            elif key == "blocked-by":
                current_task["blocked_reason"] = _extract_quoted(value)
            elif key == "note":
                current_task["note"] = _extract_quoted(value)
            elif key == "url":
                current_task["url"] = value
            elif key == "depends":
                current_task["depends_on"] = [d.strip() for d in value.split(",") if d.strip()]
            continue

    return result
